### DISCUSSION TITLE
Linking G+Smo together with other libraries that also use Eigen leads to problems

### DISCUSSION THREAD

**Discussion by osmihran** [gismo/gismo#566] [Q&A] [answered]
Hello everyone,

when I try to link G+Smo and some other library that also uses Eigen into a bigger product, bad things start happening. The rest of the text is organised as follows.

1. An example demonstrating the problem.
2. My view as to what can be done about it.

Apologies for it being long, bear with me, please. (-;

### Example ###
We link G+Smo plus a library "another" together into an executable called "main". It assumes that we have compiled G+Smo in **release** mode and installed it to **/path/to/install** folder and that there is a copy of **Eigen** in **/path/to/**. Here's a scheme:
```
/path/to
       ├── eigen
       │   ├── ...
       │   └── Eigen
       ├── gismo
       │   ├── ...
       │   └── external
       │       ├── ...
       │       └── Eigen
       ├── release
       ├── install
       │   ├── include
       │   ├── lib
       │   └── share
       ├── another
       │   ├── another.cpp
       │   ├── another.h
       │   └── Makefile
       ├── main.cpp
       └── Makefile           
```

**another/another.h**
```cpp
#include <Eigen/IterativeLinearSolvers>

void useEigen();
```

**another/another.cpp**
```cpp
#include <Eigen/IterativeLinearSolvers>

#include <iostream>

void useEigen()
{
    Eigen::SparseMatrix<double> lhs(2, 2);
    lhs.insert(0, 0) = 1;
    lhs.insert(1, 1) = 1;

    Eigen::Vector2d rhs, res;
    rhs(0) = 1;
    rhs(1) = 2;

    typename Eigen::BiCGSTAB<Eigen::SparseMatrix<double>, Eigen::IncompleteLUT<double> > solver;
    solver.compute(lhs);
    std::cout << "Solution to the system is:\n" << solver.solve(rhs) << "." << std::endl;
}
```
**another/Makefile**
```bash
CPPFLAGS += -Wall -fPIC -I/path/to/eigen

LDFLAGS += -shared

.PHONY: all clean

all: libanother.so

libanother.so: another.cpp
	$(LINK.cc) $^ -o $@ $(LDLIBS)

clean:
	@- $(RM) libanother.so
```
**main.cpp**
```cpp
#include <gismo.h>
#include <another.h>

void useGismo()
{
    gismo::gsBSplineBasis<double> basis(0.0, 1.0, 0, 2);
    gismo::gsMatrix<double> u(1, 1);
    u(0, 0) = 0.5;

    gismo::gsMatrix<double> val = basis.eval(u);
    std::cout << "3rd quadratic Bernstein polynomial at 0.5 evaluates to " << val(2, 0) << "." << std::endl;
}

int main()
{
    useGismo();
    useEigen();

    return 0;
}
```
**Makefile**
```
CPPFLAGS += -Wall -mavx2# I seem to need -mavx2 on my computer, cf. issue 374.
CPPFLAGS += -I/path/to/install/include/gismo
CPPFLAGS += -I/path/to/eigen -Ianother

LDFLAGS += -L/path/to/another
LDFLAGS += -L/path/to/install/lib
LDFLAGS += -Wl,-rpath,/path/to/another
LDFLAGS += -Wl,-rpath,/path/to/install/lib

LDLIBS += -lanother
LDLIBS += -lgismo

.PHONY: all clean

all: main

main: main.cpp
	$(LINK.cc) $^ -o $@ $(LDLIBS)

clean:
	@- $(RM) main
````
Let's compile everything:
```
> cd /path/to/another
> make
> cd ..
> make
```
Running the executable leads to
```
> ./main 
3rd quadratic Bernstein polynomial at 0.5 evaluates to 0.25.
realloc(): invalid old size
Aborted (core dumped)
```
**Exercise:** Comment out the body of the function ```useGismo()```; the program should work correctly now.

### Discussion ###
It is my understanding that the problem here is the co-existence of two copies of Eigen, which leads to a confusion during  runtime. ~~In my experience, making the "another" library use the copy of Eigen in G+Smo does not solve the problem.~~
 
#### Solution 1 ####
We can rename the namespace ```Eigen``` of Eigen in G+Smo to ```GS_eigen```. On Linux, this can be done with two calls to ```find``` and ```sed```:

```
> find . -type f -exec sed -i 's/namespace Eigen/namespace GS_eigen/g' {} +
> find . -type f -exec sed -i 's/Eigen:/GS_eigen:/g' {} +
```
After re-compiling and re-installing G+Smo as well as of the "another" library, everything works now:
```
> cd build
> make -j 8
> make install
> cd /path/to
> make
> ./main 
3rd quadratic Bernstein polynomial at 0.5 evaluates to 0.25.
Solution to the system is:
1
2.
```
#### Caveats ####
1. If you want the namespace to be called ```gsEigen``` instead of ```GS_eigen```, you have to be a bit more careful with editing, as there already are some uses of ```gsEigen``` in the source code of G+Smo.
2. For this to work the way I have shown, one has to include G+Smo _before_ including Eigen in the Makefile of main (**exercise** swap the order of ```CPPFLAGS``` in the Makefile of main to obtain a very confusing error message). A cleaner and more reliable solution would be to rename the folder **gismo/external/Eigen** to **gismo/external/gsEigen** but yet again, one has to be careful with editing the includes afterwards.

#### Solution 2 ####
If I remember correctly, @filiatra is aware of the problem and has suggested to wrap the ```Eigen``` namespace inside ```gismo```. Feel free to add details.

#### Solution 3 ####
Any ideas?

**Comment by filiatra** [comment]
Hi and thnx for this report,

I tried Solution 2 but it is rather complicated to do due to all the system file inclusions etc.

A simple Solution 3:
Just Rename external/Eigen  to external/gsEigen   and update inclusion in gsCore/gsLinearAlgebra.h
I do not think Solution 3 has the same effect as Solution 1.  It will not solve the linking issue..

There will be always a possible problem of course: As you mention in Caveat 2, we need to impose the Eigen headers of G+Smo over the Eigen headers of "another". This might have an impact on "another" in the sense that it might expect a different version of Eigen. The only complete solution I can see here is a kind of Solution 2. But it looks very hard to do.

**Comment by filiatra** [comment]
OK here is my proposed solution, finally closer to Solutions 1 and 2:

- Wrap and rename Eigen namespace as  gismo::gsEigen
- Rename external/Eigen to external/gsEigen to avoid external includes
- Undefine all eigen guards so that loading again Eigen library from another source is possible.

It seems to work. Including "gismo.h" before Eigen (g. in main.cpp) is recommended, and works fine. 

Including "gismo.h" after Eigen also works but produces some warnings:

warning: "EIGEN_DEFAULT_DENSE_INDEX_TYPE" redefined
warning: "eigen_assert" redefined
warning: "EIGEN_STATIC_ASSERT" redefined

These warn about the configuration of Eigen has changed inside G+Smo, therefore they can be useful to keep.

Here is the commit (at branch eigen_fix) : https://github.com/gismo/gismo/compare/eigen_fix

**Comment by osmihran** [comment]
> OK here is my proposed solution, finally closer to Solutions 1 and 2:
> 
>     * Wrap and rename Eigen namespace as  gismo::gsEigen
> 
>     * Rename external/Eigen to external/gsEigen to avoid external includes
> 
>     * Undefine all eigen guards so that loading again Eigen library from another source is possible.
> 
> 
> It seems to work. Including "gismo.h" before Eigen (g. in main.cpp) is recommended, and works fine.

Thanks! You are right in that we should also do something about folder naming and include guards.

> Here is the commit (at branch eigen_fix) : https://github.com/gismo/gismo/compare/eigen_fix

I took the liberty of committing two little fixes. I will keep you updated about my testing progress.

**Comment by osmihran** [comment]
After spending a weekend with this I am more confused than before. What seems clear is that there is a bug in my example: the Makefile of "another" should use the same compiler flags as the Makefile of "main", i.e., line 1 should read
```
CPPFLAGS += -Wall -mavx2 -fPIC -I/path/to/eigen
```
The Eigen problem then disappears, at least in this example. @filiatra , can you reproduce it?

**Comment by filiatra** [comment]
Thnx for the fixes.
I could reproduce your original example (got errors and after the eigen_fix it worked).I used: 
  CPPFLAGS += -Wall -fPIC
in the makefile of "another".

I think it is expected that if all compiler flags are the same and the eigen version is the same then eventually it will work even without renaming the namespace.
But if there is really a different version, possibly compiled with different flags etc, than that's the real problem to solve.

**Comment by osmihran** [comment]
Here's one more idea (suggested by a friend of mine):

**Solution 4**
Add a CMake switch to decide whether to use Eigen from G+Smo or from outside.

Say we call this switch ```I_ALREADY_HAVE_EIGEN``` and default it to ```OFF```. Normal user would not see anything and  G+Smo would behave the same way as it always did. But if some crazy person (such as me) comes with their own copy of Eigen, one would do something like
```
> cmake /path/to/gismo -DI_ALREADY_HAVE_EIGEN=ON -DEIGEN_PATH=/path/to/eigen
> make
> make install
```
to obtain an installed version of G+Smo that does not have Eigen in the install folder. One Eigen, no problem!

This has a few advantages compared to renaming the namespace:
- easier to program (should be a few lines in CMake);
- easier to maintain (no need to move 300+ files if G+Smo decides to upgrade its own copy of Eigen);
- cleaner (personal opinion based on fear of shooting myself in the foot with the ```#define```s)
- easier to understands to outsiders (Eigen as an external dependency is a common thing, cf. e.g., CGAL)

What do you think?

**Comment by filiatra** [comment]
> **Solution 4**
> Add a CMake switch to decide whether to use Eigen from G+Smo or from outside.

Yes this is a good idea and has I think we should do that. But I find it is also independent from the question to some extent:

It does not really solve your problem where you have a library already compiled with Eigen possibly on a different machine and you try to link it to gismo after compiling it (possibly with Eigen installed in the system).

So maybe Solution 3+4 should co-exist?

**Comment by osmihran** [comment]
> It does not really solve your problem where you have a library already compiled with Eigen possibly on a different machine and you try to link it to gismo after compiling it (possibly with Eigen installed in the system).

Ah, I see, you want to solve a more general problem. My goal was more modest: I was aiming at compiling both G+Smo and the other library on the same machine, with the same compiler (to avoid name-mangling problems) and with the same Eigen. In the end, I would like to be able to do something like
```
> g++ -I/path/to/eigen -isystem/path/to/other/include -isystem/path/to/gismo/include application.cpp -L/path/to/other/lib -L/path/to/gismo/lib -lother -lgismo
```
I think that Solution 4 would be enough for this but not for the situation you mention. To be honest, that situation seems quite difficult to me and I am not sure if we can realistically hope to be able solve it. Or do you think that Solution 3 would be enough then?

> So maybe Solution 3+4 should co-exist?

If we go for Solution 3, I don't see any advantage of going for 4 as well. Am I missing something?

**Comment by filiatra** [comment]
For sure the more general problem can be solved to some extent, there could always be conflicts.

The interest I see Solution 4 is two points:
- Solve the problem of co-existing Eigen when compiling on the same machine, by giving freedom to the user to choose
- When packaging for linux it would be needed to use the Eigen package installation of the linux system and not sneak in our own copy.

For Solution 3:
- Avoid conflicts when linking with other libraries that use Eigen (which does arise in practice). Both for header inclusion and for symbol name lookup.


Overall I can see situations where either delegating to external Eigen or wanting to use our own private & renamed copy will came in handy

**Comment by osmihran** [comment]
> For sure the more general problem can be solved to some extent, there could always be conflicts.
> 
> The interest I see Solution 4 is two points:
> 
>     * Solve the problem of co-existing Eigen when compiling on the same machine, by giving freedom to the user to choose
> 
>     * When packaging for linux it would be needed to use the Eigen package installation of the linux system and not sneak in our own copy.
> 
> 
> For Solution 3:
> 
>     * Avoid conflicts when linking with other libraries that use Eigen (which does arise in practice). Both for header inclusion and for symbol name lookup.

I agree with you in all the points you mention.

> Overall I can see situations where either delegating to external Eigen or wanting to use our own private & renamed copy will came in handy

Yes. For me it is difficult to imagine using Solution 3 _and_ 4 at the same time. So I guess we need to choose one of them or come up with something yet different.

Do you know of any guidelines or best practices that we can read? I have spent quite some time googling around but without much luck except for [this suggestion](https://stackoverflow.com/a/32225568/8634018) to do what we did with Eigen so far.
Surely we are not the first ones to run into this kind of problem.

**Comment by osmihran** [comment]
**Solution 5**

Make **another** depend _not_ the copy of Eigen in **/path/to/eigen** but to G+Smo's Eigen in **/path/to/gismo/external**.

This solution is surprisingly simple and has been working very well for me since 1.5 years or so. It would break though if **another** would follow the same philosophy as G+Smo, i.e., if it were shipped with its own copy of Eigen. Since this is not very probable, I think there is no acute need for action in G+Smo now.

**Comment by hverhelst** [comment]
@filiatra @osmihran shall we move this to a discussion?

**Comment by osmihran** [comment]
Yes, please.

**Comment by dominik-mokris-mtu** [comment, marked as answer]
We have gone for solution 1 in PR #617 . This still does not prevent using solution 5.
